enhancements/windows-containers/health-management.md

```diff
@@ -0,0 +1,311 @@
+
+
+## Proposal
+
+Currently, Windows Node configuration process is split between two components, WMCO and [WMCB](https://github.com/openshift/windows-machine-config-operator/).
+WMCB is a program which performs a one-shot configuration of a Windows instance, to ensure that it can become a worker
+Node.
+
+To accomplish the goals listed in this enhancement, I am proposing that WMCB be converted into a daemon,
+Windows Instance Config Daemon (WICD). WICD will have the following responsibilities:
+* WICD maintains current responsibilities of bootstrapping kubelet through the `bootstrap` command.
+* Using the configuration provided by a ConfigMap created by WMCO, WICD maintains the state of Windows Services on the instance.
+* WICD reverts all changes made to an instance when ran with the `cleanup` command, and also deletes the Node
+  object.
+
+To enable this, WMCO will have these changes in responsibility:
+* WMCO will no longer configure the Windows services on Windows instances, except for WICD.
+* WMCO will create and maintain a ConfigMap which provides WICD with the specifications for each Windows service that
+  must be created on a Windows instance.
+* WMCO will invoke the `bootstrap` command of WICD when initially adding a Windows instance to the cluster.
+* WMCO will invoke the `cleanup` command of WICD when removing an instance from the cluster. WMCO will no longer
+  delete the Node object.
+
+The Windows Node creation workflow for the end user will not be changed.
+
+### User Stories
+
+User stories can be found within the health management epic, [WINC-657](https://issues.redhat.com/browse/WINC-657).
+
+### API Extensions
+
+N/A
+
+### Risks and Mitigations
+
+* It is possible that issues with Windows instances will occur which WICD is unable to resolve, causing the Node to
+  enter a `NotReady` state. For instances created through the machine-api, the user must set up MachineHealthChecks,
+  which will correct the state by recreating the Machine. For BYOH instances, user intervention is required at this
+  point. Events and alerts will ensure that a cluster administrator is made aware of the issue.
+* There is a possibility that a cluster admin could mistakenly remove the services ConfigMap for an older WMCO version,
+  before Node upgrade is completed. This would result in WICD not knowing what services it has configured on the instance.
+  In order to prevent this, WICD can set the Windows service description to "openshift-managed", so that it
+  can search for services and remove them. In order to facilitate upgrades from a version without WICD, this string
+  should be added to all existing services in 4.10, one major version ahead of WICD's target release.
+
+## Design Details
+
+### Services ConfigMap
```

Comment:
With services like hybrid-overlay, WMCO waits for HNS networks and annotations on node objects to be present before proceeding. How will that be handled / expressed in the ConfigMap?

Reply:
I still don't understand how you will handle [waiting for HNS networks](https://github.com/openshift/windows-machine-config-operator/blob/master/pkg/windows/windows.go#L431) which is done by WMCO today.

Reply:
I understand now, that is a command that we run which does not have an output used in a Windows service.

That wait will covered as are retrying for the `distributed-router-gateway-mac` Node annotation, which is only placed on the Node [once the networks have been created](https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/hybrid-overlay/pkg/controller/node_windows.go#L309).
Additionally we are [creating an endpoint](https://github.com/openshift/windows-machine-config-operator/blob/bb5f6d48a340569fd203d2de12237df4d445fb57/pkg/windows/windows.go#L830) which requires the HNS network, so that check should be built into the PowerShell script which covers that. I'll make a note of that in the enhancement.